lib/improver/argparser.py

```diff
@@ -28,19 +28,185 @@
+    #          to the ArgumentParser.add_argument() method
+
+    # All CLIs should select something from this dictionary:
+    # NB: --help included for free with ArgumentParser
+    CENTRALIZED_ARGUMENTS = {
+        'input_file': (
+            ['input_filepath'],
+            {}),
+        'output_file': (
+            ['output_filepath'],
+            {}),
+    }
+
+    # *All* CLIs will use the options here (no option to disable them):
+    COMPULSORY_ARGUMENTS = {
+        # TODO: Implement (some of?) these:
+        #        'debug': (
+        #            ['--debug'],
+        #            {'action': 'store_true'}),
+        #        'dry_run': (
+        #            ['--dry-run'],
+        #            {'action': 'store_true'}),
+        #        'profile' : (
+        #            ['--profile'],
+        #            {'action': 'store_true'}),
+        #        'verbose': (
+        #            ['--verbose', '-v'],
+        #            {'action': 'count'}),
+        #        'version': (
+        #            ['--version'],
+        #            {'action': 'version',
+        #             'version': improver.version_message() }),
+    }
+
+    # We can override including these, but options common to everything should
+    # be in a list here:
+    # DEFAULT_CENTRALIZED_ARGUMENTS = ('input_file', 'output_file')
+    DEFAULT_CENTRALIZED_ARGUMENTS = ()
+
+    def __init__(self, central_arguments=DEFAULT_CENTRALIZED_ARGUMENTS,
+                 specific_arguments=None, **kwargs):
+        """Create an ArgParse instance, which is a subclass of
+        argparse.ArgumentParser and automatically add all of the arguments.
+        (Note: The ArgParse.COMPULSORY_ARGUMENTS are always added.)
+
+        Args:
+            central_arguments (list):
+              A list containing the centralized arguments we require.
+              (Keys of the centralized argument dictionary). By default this is
+              set as ArgParse.DEFAULT_CENTRALIZED_ARGUMENTS.
+            specfic_arguments (list):
+              A list of argument specifications required to add arguments
+              which are not contained within the centralized argument
+              dictionary. The format of these argument specifications should
+              be the same as the values in the centralized argument dictionary.
+              (For more details, see the add_arguments method).
+              Default is None, which does not add additional arguments.
+            kwargs (dictionary):
+              Additional keyword arguments which are passed to the superclass
+              constructor (argparse.ArgumentParser).
+        """
+
+        self._args = None
+
+        # Allow either central_arguments or specific_arguments to be None
+        # (or empty lists)
+        if central_arguments is None:
+            central_arguments = []
+        if specific_arguments is None:
+            specific_arguments = []
```

Comment:
Would it be simpler to allow these to default to empty lists?  Don't think this is needed here.

Reply:
It's a bad idea to have a mutable variable as a default argument to a function/method. This approach is considered better practice.
See: https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

Reply:
It is buried in the Google one: https://google.github.io/styleguide/pyguide.html#Default_Argument_Values